src/plugins/factoids/storage.internal.js

```diff
@@ -0,0 +1,171 @@
+/* eslint-disable no-underscore-dangle */
+const fs = require('fs');
+const path = require('path');
+const { promisify, inspect } = require('util');
+
+const readFile = promisify(fs.readFile);
+const writeFile = promisify(fs.writeFile);
+
+const FILE_PATH = path.join(__dirname, 'factoids.json');
+
+function toKey(key) {
+  return (key && key.toLowerCase()) || null;
+}
+
+class Store {
+  constructor() {
+    this._items = new Map();
+    this._loaded = false;
+    this._readPromise = null;
+    this._writePromise = null;
+    this._dirty = false;
+  }
+
+  _replaceAll(items) {
+    this._items = items;
+  }
+
+  set(key, entry) {
+    this._items.set(key.toLowerCase(), entry);
+    this._dirty = true;
+    return this;
+  }
+
+  getText(key) {
+    const MAX_ALIAS_DEPTH = 5;
+    let cursor = { type: 'alias', value: toKey(key) };
+
+    for (
+      let i = 0;
+      i < MAX_ALIAS_DEPTH && cursor && cursor.type === 'alias';
+      i += 1
+    ) {
+      cursor = this._items.get(cursor.value) || null;
+    }
+
+    if (cursor && cursor.type === 'alias') {
+      throw new Error(`Alias depth exceeded when looking up a factoid.`);
+    }
+
+    if (cursor && cursor.type === 'factoid') {
+      return cursor.value;
+    }
+
+    return null;
+  }
+
+  toJSON() {
+    const obj = {};
+    for (const [key, entry] of this._items) {
+      obj[key] = entry;
+    }
+    return obj;
+  }
+
+  needsLoadFromDisk() {
+    return !this._loaded;
+  }
+
+  async loadFromDisk() {
+    const readPromise = this._readPromise || this._loadFromDiskInternal();
+    this._readPromise = readPromise;
+
+    await readPromise;
+
+    this._loaded = true;
+    if (this._readPromise === readPromise) {
+      this._readPromise = null;
+    }
+  }
+
+  update(key, { editor, value }) {
+    if (!this._loaded) {
+      throw new Error(
+        `Must wait for loadFromDisk to complete before calling .update`,
+      );
+    }
+
+    if (value.length > 400) {
```

Comment:
In IRC, what is the message limit actually measured in? Is it bytes? Might need to correct this.

Reply:
is it hardcoded in the protocol, or determined by the server?

Reply:
in node-irc and splitlong.pl, the max length is calculated as

497 - nick length - hostmask length

the protocol gives a limit of 512 but things like `PRIVMSG` decrease this number

Reply:
Thanks for clarifying. In that case, we should do utf-16 string -> utf8 byte length and check, right?

Reply:
ah, hadn't realised this - have just been checking the JS string length

I doubt the edge cases that affect it will crop up in factoids, but it definitely makes sense to do so.

using this to calculate it now;

```function byteCount(s) {
    return encodeURI(s).split(/%..|./).length - 1;
}```